From 34264097dc6fdbe25057859aadeb2037828e8d06 Mon Sep 17 00:00:00 2001
From: Mike Snitzer <snitzer@redhat.com>
Date: Fri, 13 Jun 2014 14:36:33 -0400
Subject: [block] blk-mq: fix regression from commit 624dbe475416

Message-id: <1402670238-13916-106-git-send-email-snitzer@redhat.com>
Patchwork-id: 83878
O-Subject: [RHEL7.1 PATCH 105/150] blk-mq: fix regression from commit 624dbe475416
Bugzilla: 1105204
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Jeff Moyer <jmoyer@redhat.com>

BZ: 1105204

Upstream commit f899fed4421d6b098ed6a9d69303c70e590bf2c0
Author: Jens Axboe <axboe@fb.com>
Date:   Wed Jun 4 09:11:53 2014 -0600

    blk-mq: fix regression from commit 624dbe475416

    When the code was collapsed to avoid duplication, the recent patch
    for ensuring that a queue is idled before free was dropped, which was
    added by commit 19c5d84f14d2.

    Add back the blk_mq_tag_idle(), to ensure we don't leak a reference
    to an active queue when it is freed.

    Signed-off-by: Jens Axboe <axboe@fb.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/block/blk-mq.c b/block/blk-mq.c
index 96e6eb6..4e8e8cf 100644
--- a/block/blk-mq.c
+++ b/block/blk-mq.c
@@ -1552,6 +1552,8 @@ static void blk_mq_exit_hw_queues(struct request_queue *q,
   if (i == nr_queue)
    break;
 
+  blk_mq_tag_idle(hctx);
+
   if (set->ops->exit_hctx)
    set->ops->exit_hctx(hctx, i);
 
-- 
1.7.1